synapse/federation/federation_client.py

```diff
@@ -411,43 +426,40 @@ async def get_pdu(
                         room_version=room_version,
                         timeout=timeout,
                     )
+                    pull_origin = destination
 
                     pdu_attempts[destination] = now
 
                     if event:
                         # Prime the cache
-                        self._get_pdu_cache[event.event_id] = event
+                        self._get_pdu_cache[event.event_id] = (event, pull_origin)
 
                         # Now that we have an event, we can break out of this
                         # loop and stop asking other destinations.
                         break
 
-                except SynapseError as e:
+                except NotRetryingDestination as e:
+                    logger.info(f"get_pdu(event_id={event_id}): {e}")
+                    continue
+                except FederationDeniedError:
                     logger.info(
-                        "Failed to get PDU %s from %s because %s",
-                        event_id,
-                        destination,
-                        e,
+                        f"get_pdu(event_id={event_id}): Not attempting to fetch PDU from {destination} because the homeserver is not on our federation whitelist"
```

Comment:
Consistently handling these two errors across `get_pdu` , backfill, and the generic `_try_destination_list` which is used for many places we use this pattern.